src/core/server/saved_objects/service/lib/repository.ts

```diff
@@ -258,6 +261,15 @@ export class SavedObjectsRepository {
       }
     }
 
+    if (
+      'credential' === type &&
+      typeof attributes !== 'undefined' &&
+      typeof attributes.password !== 'undefined'
+    ) {
+      attributes.password = await this._crypto_cli.encrypt(attributes.password);
+    }
```

Comment:
why do we add encryption logic a save object service?
saved object service should be low level data access api without adding too much specific business logic.

credential is one of saved object type, the implementation should not touch code here

Reply:
> fail to run `yarn osd bootstrap`
> 
> ```
> success Saved lockfile.
> 
>  succ yarn.lock analysis completed without any issues
>  succ 22 bootstrap builds are cached
>  info [opensearch-dashboards] running [osd:bootstrap] script
> ERROR [bootstrap] failed:
> ERROR Error: Command failed with exit code 1: /Users/huanji/.nvm/versions/node/v14.19.1/lib/node_modules/yarn/bin/yarn.js run osd:bootstrap
>       error Command failed with exit code 1.
>       $ node scripts/build_ts_refs && node scripts/register_git_hook
>       ERROR Error: Command failed with exit code 1: /Users/huanji/src/huan-osd/node_modules/typescript/bin/tsc -b tsconfig.refs.json --pretty
>             src/core/server/saved_objects/service/lib/crypto/crypto_cli.ts:42:8 - error TS6307: File '/Users/huanji/src/huan-osd/src/core/server/saved_objects/service/lib/crypto/crypto_metadata.json' is not listed within the file list of project '/Users/huanji/src/huan-osd/src/core/tsconfig.json'. Projects must list all files or use an 'include' pattern.
> 
>             42 } from './crypto_metadata.json';
>                       ~~~~~~~~~~~~~~~~~~~~~~~~
> 
>             src/core/server/saved_objects/service/lib/repository.ts:267:25 - error TS2339: Property 'password' does not exist on type 'T'.
> 
>             267       typeof attributes.password !== 'undefined'
>                                         ~~~~~~~~
> 
>             src/core/server/saved_objects/service/lib/repository.ts:269:18 - error TS2339: Property 'password' does not exist on type 'T'.
> 
>             269       attributes.password = await this._crypto_cli.encrypt(attributes.password);
>                                  ~~~~~~~~
> 
>             src/core/server/saved_objects/service/lib/repository.ts:269:71 - error TS2339: Property 'password' does not exist on type 'T'.
> 
>             269       attributes.password = await this._crypto_cli.encrypt(attributes.password);
>                                                                                       ~~~~~~~~
> 
> 
>             Found 4 errors.
> 
>                 at makeError (/Users/huanji/src/huan-osd/node_modules/execa/lib/error.js:59:11)
>                 at handlePromise (/Users/huanji/src/huan-osd/node_modules/execa/index.js:114:26)
>                 at processTicksAndRejections (internal/process/task_queues.js:95:5)
>                 at buildRefs (/Users/huanji/src/huan-osd/src/dev/typescript/build_refs.ts:41:5)
>                 at buildAllRefs (/Users/huanji/src/huan-osd/src/dev/typescript/build_refs.ts:35:3)
>                 at description (/Users/huanji/src/huan-osd/src/dev/typescript/build_refs.ts:51:7)
>                 at /Users/huanji/src/huan-osd/packages/osd-dev-utils/target/run/run.js:64:13
>                 at Object.withProcRunner (/Users/huanji/src/huan-osd/packages/osd-dev-utils/target/proc_runner/with_proc_runner.js:45:9)
>                 at run (/Users/huanji/src/huan-osd/packages/osd-dev-utils/target/run/run.js:63:9)
>       info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
>           at makeError (/Users/huanji/src/huan-osd/packages/osd-pm/dist/index.js:25150:11)
>           at handlePromise (/Users/huanji/src/huan-osd/packages/osd-pm/dist/index.js:24085:26)
>           at processTicksAndRejections (internal/process/task_queues.js:95:5)
>           at async /Users/huanji/src/huan-osd/packages/osd-pm/dist/index.js:9051:9
>           at async scheduleItem (/Users/huanji/src/huan-osd/packages/osd-pm/dist/index.js:10938:9)
> error Command failed with exit code 1.
> info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
> ```

This was not visible when submitting the PR, but based on the error message, it is related to dynamic type casting.

> why do we add encryption logic a save object service? saved object service should be low level data access api without adding too much specific business logic.
> 
> credential is one of saved object type, the implementation should not touch code here

For the time being, it is hardcoded as part of the proof of concept. Will refactor it after learning more about the codebase. Before that, do you have any suggestions or proposals?

Reply:
It is not about hard code. it is about it should not adding any credential specific logic to data access tier. Credential type is only one of Saved Object type.

current poc appraoch is against [Dependency inversion principle] https://en.wikipedia.org/wiki/Dependency_inversion_principle)

Savedobject is low-level module, credential management is high-level module
`high-level modules should independent of the low-level module implementation details`

One possible approach should be.
Browser --> invoke Credential RESTFul API through http --> Credential API handle receive the request --> Invoke Saved Object API

Reply:
let's do some abstraction of credentials, we will support other auth types in addition to basic auth in the future